<!DOCTYPE html>
<html lang="en">
{% extends "snpdb/dashboard.html" %}


{% block content %}
{% load display_rows %}

<div class="container" style="padding-bottom:30px">
    <div class="page-header text-header" style="padding-left:30px">
        <h1>Difference Between Libraries</h1>
        <h4>There are {{ count.snp_count }} snps found in {{library1}} that are not found in {{library2}}. </h4>
    </div>
</div>

<body>
<div class="container-fluid" style="padding-bottom:30px">
    <div class="collapse navbar-collapse navbar-ex1-collapse">
        <ul class="nav navbar-nav">
            <form class="form-inline" role="form" id='search' action='/snpdb/difference_two_libraries/search' method="get">
                <div class="form-group">
                    <label class="sr-only" for="library">Library 1</label>
                    <input type="text" class="form-control" name="lib1" placeholder="library 1">
                </div>
                <div class="form-group">
                    <label class="sr-only" for="library">Library 2</label>
                    <input type="text" class="form-control" name="lib2" placeholder="library 2">
                </div>
                <select class="form-control" name="f" label="Effect Impact">
                    <option value>-- Select Effect Impact--</option>
                    <option value="MODIFIER">Modifier</option>
                    <option value="LOW">Low Impact</option>
                    <option value="MODERATE">Moderate Impact</option>
                    <option value="HIGH">High Impact</option>
                    <option value="all">All</option>
                </select>
                <button id="submit" class="btn btn-primary" type="submit">Search</button>
            </form>
        </ul>
    </div>
</div>
</body>

<div class="container-fluid" style="padding-bottom:30px">
    <div class="row-fluid">
        <div class="col-md-6">
            <div class="panel panel-primary">
                <div class="panel-heading"><a href="/snpdb/impact-snps/search?lib1={{library1}}&lib2={{library2}}&count={{high_count}}&impact=HIGH" style="color: rgb(255,255,255)">High Impact</a></div>
                <ul class="list-group">
                    {% for x in snps%}
                        {% if x.effect__effect_string == "HIGH" %}
                            <li class="list-group-item">Number of Snps: {{ x.snp_count }}</li>
                        {% endif %}
                    {% endfor %}
                    <li class="list-group-item">Number of genes: {{gene|get_at_index:0}}</li>
                    {% for y in impacts %}
                        {%if y.effect__effect_string == "HIGH" %}
                            <li class="list-group-item">Number of {{y.effect__effect_class}}: {{ y.dcount }}</li>
                        {% endif %}
                    {% endfor %}
                </ul>
            </div>
        </div>
        <div class="col-md-6">
            <div class="panel panel-primary">
                <div class="panel-heading"><a href="/snpdb/impact-snps/search?lib1={{library1}}&lib2={{library2}}&count={{low_count}}&impact=LOW" style="color: rgb(255,255,255)">Low Impact</a></div>
                <ul class="list-group">
                    {% for x in snps%}
                        {% if x.effect__effect_string == "LOW" %}
                            <li class="list-group-item">Number of Snps: {{ x.snp_count }}</li>
                        {% endif %}
                    {% endfor %}
                            <li class="list-group-item">Number of genes: {{gene|get_at_index:2}}</li>
                    {% for y in impacts %}
                        {%if y.effect__effect_string == "LOW" %}
                            <li class="list-group-item">Number of {{y.effect__effect_class}}: {{ y.dcount }}</li>
                        {% endif %}
                    {% endfor %}
                </ul>
            </div>
        </div>
    </div>
    <div class="row-fluid">
        <div class="col-md-6">
            <div class="panel panel-primary">
                <div class="panel-heading"><a href="/snpdb/impact-snps/search?lib1={{library1}}&lib2={{library2}}&count={{moderate_count}}&impact=MODERATE" style="color: rgb(255,255,255)">Moderate Impact</a></div>
                <ul class="list-group">
                    {% for x in snps%}
                        {% if x.effect__effect_string == "MODERATE" %}
                            <li class="list-group-item">Number of Snps: {{ x.snp_count }}</li>
                        {% endif %}
                    {% endfor %}
                    <li class="list-group-item">Number of genes: {{gene|get_at_index:1}}</li>
                    {% for y in impacts %}
                        {%if y.effect__effect_string == "MODERATE" %}
                            <li class="list-group-item">Number of {{y.effect__effect_class}}: {{ y.dcount }}</li>
                        {% endif %}
                    {% endfor %}
                </ul>
            </div>
        </div>
        <div class="col-md-6">
            <div class="panel panel-primary">
                <div class="panel-heading"><a href="/snpdb/impact-snps/search?lib1={{library1}}&lib2={{library2}}&count={{modifier_count}}&impact=MODIFIER" style="color: rgb(255,255,255)">Modifier Impact</a></div>
                <ul class="list-group">
                    {% for x in snps%}
                        {% if x.effect__effect_string == "MODIFIER" %}
                            <li class="list-group-item">Number of Snps: {{ x.snp_count }}</li>
                        {% endif %}
                    {% endfor %}
                    <li class="list-group-item">Number of genes: {{gene|get_at_index:3}}</li>
                    {% for y in impacts %}
                        {%if y.effect__effect_string == "MODIFIER" %}
                            <li class="list-group-item">Number of {{y.effect__effect_class}}: {{ y.dcount }}</li>
                        {% endif %}
                    {% endfor %}
                </ul>
            </div>
        </div>
    </div>
</div>



<div class="container-fluid" style="padding-bottom:30px">
    <div class="form-group">
        <div class="alert alert-info">
            <p class="alert-info">Further information about Effect Impacts can be found at <a href="http://snpeff.sourceforge.net/SnpEff_manual.html#output">SnpEff Manual</a> or by expanding the boxes below.</p>
        </div>
    </div>
</div>

<div class="container-fluid" style="padding-bottom:30px">

<div class="panel-group" id="accordion2">
    <div class="panel panel-default">
        <div class="panel-heading">
            <h4 class="panel-title">
                <a data-toggle="collapse" data-parent="#accordion2" href="#collapseTwo">Modifier Examples</a>
            </h4>
        </div>
    </div>
    <div id="collapseTwo" class="panel-collapse collapse">
        <div class="panel-body">
            <table class="table table-striped">
                <tr> <th> Effect<br>Seq. Ontology </th><th> Effect<br>Classic </th><th> Note &amp; Example </th><th>Impact</th> </tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001580"> coding_sequence_variant </a>  </td><td> CDS  </td><td> The variant hits a CDS. </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001576"> transcript_variant </a>  </td><td> TRANSCRIPT  </td><td> The variant hits a transcript. </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001566"> regulatory_region_variant </a>  </td><td> REGULATION  </td><td> The variant hits a known regulatory feature (non-coding). </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001631"> upstream_gene_variant </a>  </td><td> UPSTREAM  </td><td> Upstream of a gene (default length: 5K bases) </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001624"> 3_prime_UTR_variant </a>  </td><td> UTR_3_PRIME  </td><td> Variant hits 3'UTR region </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001632"> downstream_gene_variant </a>  </td><td> DOWNSTREAM  </td><td> Downstream of a gene (default length: 5K bases) </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001564"> gene_variant </a>  </td><td> GENE  </td><td> The variant hits a gene. </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000605"> intergenic_region</a> </td><td> INTERGENIC </td><td> The variant is in an intergenic region </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002017"> conserved_intergenic_variant</a> </td><td> INTERGENIC_CONSERVED </td><td> The variant is in a highly conserved intergenic region </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002011"> intragenic_variant</a> </td><td> INTRAGENIC </td><td> The variant hits a gene, but no transcripts within the gene</td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001627"> intron_variant </a>  </td><td> INTRON  </td><td> Variant hits and intron. Technically, hits no exon in the transcript. </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002018"> conserved_intron_variant </a>  </td><td> INTRON_CONSERVED  </td><td> The variant is in a highly conserved intronic region </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0000276"> miRNA </a>  </td><td> MICRO_RNA  </td><td> Variant affects an miRNA </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001791"> exon_variant </a>  </td><td> EXON  </td><td> The vairant hits an exon. </td> <td> MODIFIER </td></tr>
                <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001623"> 5_prime_UTR_variant </a>  </td><td> UTR_5_PRIME  </td><td> Variant hits 5'UTR region </td> <td> MODIFIER </td></tr>
            </table>
        </div>
    </div>
    </div>
    <div class="panel-group" id="accordion3">
        <div class="panel panel-default">
            <div class="panel-heading">
                <h4 class="panel-title">
                    <a data-toggle="collapse" data-parent="#libraries" href="#collapseThree">Low Impact Examples</a>
                </h4>
            </div>
        </div>
        <div id="collapseThree" class="panel-collapse collapse">
            <div class="panel-body">
                <table class="table table-striped">
                    <tr> <th> Effect<br>Seq. Ontology </th><th> Effect<br>Classic </th><th> Note &amp; Example </th><th>Impact</th> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001819"> synonymous_variant </a>  </td><td> SYNONYMOUS_CODING  </td><td> Variant causes a codon that produces the same amino acid <br>e.g.:  Ttg/Ctg, L/L </td> <td> LOW </td> </tr>
                    <tr> <td> <a href=""> start_retained </a>  </td><td> SYNONYMOUS_START  </td><td> Variant causes start codon to be mutated into another start codon. <br>e.g.:  Ttg/Ctg, L/L (TTG and CTG can be START codons) </td> <td> LOW </td> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001567"> stop_retained_variant </a>  </td><td> SYNONYMOUS_STOP  </td><td> Variant causes stop codon to be mutated into another stop codon. <br>e.g.: taA/taG, */* </td> <td> LOW </td> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001988"> 5_prime_UTR_premature start_codon_gain_variant </a>  </td><td> START_GAINED  </td><td> A variant in 5'UTR region produces a three base sequence that can be a START codon. </td> <td> LOW </td> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630"> splice_region_variant </a>  </td><td> SPLICE_SITE_BRANCH  </td><td> A varaint affective putative (Lariat) branch point, located in the intron.</td> <td> LOW </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001582"> initiator_codon_variant </a>  </td><td> NON_SYNONYMOUS_START  </td><td> Variant causes start codon to be mutated into another start codon (the new codon produces a different AA). <br>e.g.: Atg/Ctg, M/L (ATG and CTG can be START codons) </td> <td> LOW </td> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001567"> stop_retained_variant </a>  </td><td> NON_SYNONYMOUS_STOP  </td><td> Variant causes stop codon to be mutated into another stop codon (the new codon produces a different AA). <br>e.g.: Atg/Ctg, M/L (ATG and CTG can be START codons) </td> <td> LOW </td> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630"> splice_region_variant </a>  </td><td> SPLICE_SITE_REGION  </td><td> A sequence variant in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8 bases of the intron.</td> <td> LOW </td> </tr>
                </table>
            </div>
        </div>
    </div>
    <div class="panel-group" id="accordion4">
        <div class="panel panel-default">
            <div class="panel-heading">
                <h4 class="panel-title">
                    <a data-toggle="collapse" data-parent="#libraries" href="#collapseFour">Moderate Impact Examples</a>
                </h4>
            </div>
        </div>
        <div id="collapseFour" class="panel-collapse collapse">
            <div class="panel-body">
                <table class="table table-striped">
                    <tr> <th> Effect<br>Seq. Ontology </th><th> Effect<br>Classic </th><th> Note &amp; Example </th><th>Impact</th> </tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001580"> coding_sequence_variant </a>  </td><td> CODON_CHANGE  </td><td> One or many codons are changed <br>e.g.:  An MNP of size multiple of 3 </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001821"> inframe_insertion </a>  </td><td> CODON_INSERTION  </td><td> One or many codons are inserted <br>e.g.:  An insert multiple of three in a codon boundary </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001824"> disruptive_inframe_insertion</a>  </td><td> CODON_CHANGE_PLUS CODON_INSERTION  </td><td> One codon is changed and one or many codons are inserted <br>e.g.:  An insert of size multiple of three, not at codon boundary </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001822"> inframe_deletion </a>  </td><td> CODON_DELETION  </td><td> One or many codons are deleted <br>e.g.:  A deletion multiple of three at codon boundary </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001826"> disruptive_inframe_deletion </a>  </td><td> CODON_CHANGE_PLUS CODON_DELETION  </td><td> One codon is changed and one or more codons are deleted <br>e.g.:  A deletion of size multiple of three, not at codon boundary </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001583"> missense_variant </a>  </td><td> NON_SYNONYMOUS_CODING  </td><td> Variant causes a codon that produces a different amino acid <br>e.g.:  Tgg/Cgg, W/R </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001630"> splice_region_variant </a>  </td><td> SPLICE_SITE_BRANCH_U12  </td><td> A varaint affective putative (Lariat) branch point from U12 splicing machinery, located in the intron.</td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002015"> 3_prime_UTR_truncation</a> + <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572"> exon_loss </a>  </td><td> UTR_3_DELETED  </td><td> The variant deletes an exon which is in the 3'UTR of the transcript </td> <td> MODERATE </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002013"> 5_prime_UTR_truncation</a> + <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572">exon_loss_variant</a> </td><td> UTR_5_DELETED  </td><td> The variant deletes an exon which is in the 5'UTR of the transcript </td> <td> MODERATE </td></tr>
                </table>
            </div>
        </div>
    </div>
    <div class="panel-group" id="accordion5">
        <div class="panel panel-default">
            <div class="panel-heading">
                <h4 class="panel-title">
                    <a data-toggle="collapse" data-parent="#libraries" href="#collapseFive">High Impact Examples</a>
                </h4>
            </div>
        </div>
        <div id="collapseFive" class="panel-collapse collapse">
            <div class="panel-body">
                <table class="table table-striped">
                    <tr> <th> Effect<br>Seq. Ontology </th><th> Effect<br>Classic </th><th> Note &amp; Example </th><th>Impact</th> </tr>
                    <tr> <td> chromosome  </td><td> CHROMOSOME_LARGE DELETION  </td><td> A large parte (over 1%) of the chromosome was deleted. </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002008">rare_amino_acid_variant</a> </td><td> RARE_AMINO_ACID </td><td> The variant hits a rare amino acid thus is likely to produce protein loss of function </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001574"> splice_acceptor_variant </a>  </td><td> SPLICE_SITE_ACCEPTOR  </td><td> The variant hits a splice acceptor site (defined as two bases before exon start, except for the first exon). </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001575"> splice_donor_variant </a>  </td><td> SPLICE_SITE_DONOR  </td><td> The variant hits a Splice donor site (defined as two bases after coding exon end, except for the last exon). </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001578"> stop_lost </a>  </td><td> STOP_LOST  </td><td> Variant causes stop codon to be mutated into a non-stop codon <br>e.g.: Tga/Cga, */R </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0002012"> start_lost</a> </td><td> START_LOST   </td><td> Variant causes start codon to be mutated into a non-start codon. <br> e.g.: aTg/aGg, M/R </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001587"> stop_gained </a>  </td><td> STOP_GAINED  </td><td> Variant causes a STOP codon <br>e.g.: Cag/Tag, Q/* </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001572"> exon_loss_variant </a>  </td><td> EXON_DELETED  </td><td> A deletion removes the whole exon. </td> <td> HIGH </td></tr>
                    <tr> <td> <a href="http://www.sequenceontology.org/browser/current_svn/term/SO:0001589"> frameshift_variant </a>  </td><td> FRAME_SHIFT  </td><td> Insertion or deletion causes a frame shift <br>e.g.:  An indel size is not multple of 3 </td> <td> HIGH </td></tr>
                </table>
            </div>
        </div>
    </div>
</div>

{% endblock %}
</html>